library/src-bootstrapped/scala/quoted/util/ExprMap.scala

```diff
@@ -0,0 +1,158 @@
+package scala.quoted.util
+
+import scala.quoted._
+
+trait ExprMap {
+
+  /** Map an expression `e` with a type `tpe` */
+  def map[T](e: Expr[T])(given qctx: QuoteContext, tpe: Type[T]): Expr[T]
+
+  /** Map subexpressions an expression `e` with a type `tpe` */
+  def mapChildren[T](e: Expr[T])(given qctx: QuoteContext, tpe: Type[T]): Expr[T] = {
+    import qctx.tasty.{_, given}
+    class MapChildren() {
+
+      def transformStatement(tree: Statement)(given ctx: Context): Statement = {
+        def localCtx(definition: Definition): Context = definition.symbol.localContext
+        tree match {
+          case tree: Term =>
+            transformTerm(tree, defn.AnyType)
+          case tree: Definition =>
+            transformDefinition(tree)
+          case tree: Import =>
+            tree
+        }
+      }
+
+      def transformDefinition(tree: Definition)(given ctx: Context): Definition = {
+        def localCtx(definition: Definition): Context = definition.symbol.localContext
+        tree match {
+          case tree: ValDef =>
+            implicit val ctx = localCtx(tree)
+            val rhs1 = tree.rhs.map(x => transformTerm(x, tree.tpt.tpe))
+            ValDef.copy(tree)(tree.name, tree.tpt, rhs1)
+          case tree: DefDef =>
+            implicit val ctx = localCtx(tree)
+            DefDef.copy(tree)(tree.name, tree.typeParams, tree.paramss, tree.returnTpt, tree.rhs.map(x => transformTerm(x, tree.returnTpt.tpe)))
+          case tree: TypeDef =>
+            tree
+          case tree: ClassDef =>
+            ClassDef.copy(tree)(tree.name, tree.constructor, tree.parents, tree.derived, tree.self, tree.body)
+        }
+      }
+
+      def transformTermChildren(tree: Term, tpe: Type)(given ctx: Context): Term = tree match {
+        case Ident(name) =>
+          tree
+        case Select(qualifier, name) =>
+          Select.copy(tree)(transformTerm(qualifier, qualifier.tpe.widen), name)
```

Comment:
It's unclear why `qualifier.tpe.widen` is the right thing to do.

Reply:
What I need here is to have the least precise type on which this symbol can be called. Maybe I should have a look at the owner of the symbol.

Reply:
Removed